api/v1beta1/spannerautoscaleschedule_webhook.go

```diff
@@ -0,0 +1,144 @@
+	}
+
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "spanner.mercari.com", Kind: "SpannerAutoscaleSchedule"},
+		r.Name, allErrs)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *SpannerAutoscaleSchedule) ValidateUpdate(old runtime.Object) error {
+	spannerautoscaleschedulelog.Info("validate update", "name", r.Name)
+	spannerautoscaleschedulelog.V(1).Info("validating updates to SpannerAutoscaleSchedule resource", "name", r.Name, "resource", r)
+
+	var allErrs field.ErrorList
+	oldResource := old.(*SpannerAutoscaleSchedule)
+
+	// TargetResource should not be changed after creation
+	if oldResource.Spec.TargetResource != r.Spec.TargetResource {
+		err := field.Invalid(
+			field.NewPath("spec").Child("targetResource"),
+			r.Spec.TargetResource,
+			"'targetResource' can not be changed after resource has been created")
+		allErrs = append(allErrs, err)
+	}
+
+	// Disable schedule updates until reconciler is modified to propagate this change to the actual cronjobs
+	if oldResource.Spec.Schedule != r.Spec.Schedule {
+		err := field.Invalid(
+			field.NewPath("spec").Child("schedule"),
+			r.Spec.Schedule,
+			"'schedule' can not be changed after resource has been created")
+		allErrs = append(allErrs, err)
+	}
+
+	allErrs = append(allErrs, r.validateSchedule()...)
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "spanner.mercari.com", Kind: "SpannerAutoscaleSchedule"},
+		r.Name, allErrs)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *SpannerAutoscaleSchedule) ValidateDelete() error {
+	spannerautoscaleschedulelog.Info("validate delete", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object deletion.
+	return nil
+}
+
+func (r *SpannerAutoscaleSchedule) validateSchedule() field.ErrorList {
+	var allErrs field.ErrorList
+	cronSchedule, err := cronpkg.ParseStandard(r.Spec.Schedule.Cron)
+	if err != nil {
+		fldErr := field.Invalid(
+			field.NewPath("spec").Child("schedule").Child("cron"),
+			r.Spec.Schedule.Cron,
+			err.Error())
+		allErrs = append(allErrs, fldErr)
+	}
+
+	duration, err := time.ParseDuration(r.Spec.Schedule.Duration)
+	if err != nil {
+		fldErr := field.Invalid(
+			field.NewPath("spec").Child("schedule").Child("duration"),
+			r.Spec.Schedule.Duration,
+			err.Error())
+		allErrs = append(allErrs, fldErr)
+	}
+
+	// return early if either the cron schedule or the duration can not be parsed
+	if len(allErrs) != 0 {
+		return allErrs
+	}
+
+	now := time.Now()
+	firstRun := cronSchedule.Next(now)
+	secondRun := cronSchedule.Next(firstRun)
+
+	if secondRun.Before(firstRun.Add(duration)) {
+		fldErr := field.Invalid(
+			field.NewPath("spec").Child("schedule").Child("duration"),
+			r.Spec.Schedule.Duration,
+			"'duration' is larger than the time period between cron jobs of 'schedule'")
+		allErrs = append(allErrs, fldErr)
+	}
```

Comment:
I faced the following error:

```
`The SpannerAutoscaleSchedule "another-sample" is invalid: spec.schedule.duration: Invalid value: "15m30s": 'duration' is larger than the time period between cron jobs of 'schedule'`
```

This error occurred under the following conditions:

- Occurred time: 2022/04/19 11:34:47
- Schedule: `*/25 * * * *`

This is because `*/25 * * * *` means scheduling every 0, 25, 50 minutes.
In this case, the next schedule after 2022/04/19 11:34:47 is 2020/04/19 11:50:00(firstRun), and the next schedule after that is 2020/04/19 12:00:00(secondRun).

`secondRun - firstRun = 10 minutes < duration 15m30s`

So I believe that validation may be difficult given a more complex schedule than this.
FYI: https://go.dev/play/p/r1ZI8V2cBYN

Reply:
Good catch @tkuchiki san,

In addition to this, the validation does not hold good for non-periodic schedules as well (like `10 4,8,9 * * *`  for duration of `2h`). So I just kept it simple to check for just first two iterations, but now I see even that can fail sometimes.
I have now removed this validation since it is not useful.

Removing this validaiton does not have any impact on the functionality though. In the worst case, if a user provides a schedule like `0 4,5 * * *`, with duration `1h15m`, then this schedule will remain active for longer time than expected (total `2h15m`). The number of additional PUs will not increase each time a schedule is triggered, only the time will be extended. [[Ref](https://github.com/mercari/spanner-autoscaler/blob/97f0442fd5ecd65c78aa2cde325f3d9130921417/internal/scheduler/scheduler.go#L158-L169)]

Thank you for looking at it thoroughly.